src/Avalonia.Visuals/Rendering/DirtyVisuals.cs

```diff
@@ -89,19 +100,69 @@ public bool Remove(IVisual visual)
         /// <returns>A collection of visuals.</returns>
         public IEnumerator<IVisual> GetEnumerator()
         {
-            foreach (var i in _inner)
+            using (DeferChanges())
             {
-                foreach (var j in i.Value)
+                foreach (var i in _inner)
                 {
-                    yield return j;
+                    foreach (var j in i.Value)
+                    {
+                        yield return j;
+                    }
                 }
             }
         }
 
+        private DeferDisposer DeferChanges()
+        {
+            _deferring = true;
+            return new DeferDisposer(this);
+        }
+
+        private void EndDefer()
+        {
+            if (!_deferring) return;
+
+            _deferring = false;
+
+            foreach (var change in _deferredChanges)
+            {
+                switch (change.change)
+                {
+                    case DeferredChange.Add:
+                        Add(change.visual);
+                        break;
+                    case DeferredChange.Remove:
+                        Remove(change.visual);
+                        break;
+                    case DeferredChange.Clear:
+                        Clear();
+                        break;
+                }
+            }
+
+            _deferredChanges.Clear();
+        }
+
         /// <summary>
         /// Gets the dirty visuals, in ascending order of distance to their root.
         /// </summary>
         /// <returns>A collection of visuals.</returns>
         IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        private struct DeferDisposer : IDisposable
+        {
+            private DirtyVisuals _parent;
+
+            internal DeferDisposer(DirtyVisuals parent) => _parent = parent;
+
+            public void Dispose() => _parent?.EndDefer();
+        }
+
+        private enum DeferredChange
```

Comment:
If we make `Clear` an exception while enumerating and remove the `Remove` method then this enum can go away.